keanu-project/src/main/java/io/improbable/keanu/vertices/VertexLabel.java

```diff
@@ -2,24 +2,34 @@
 
 import com.google.common.base.Joiner;
 import com.google.common.collect.ImmutableList;
-import com.google.common.collect.Lists;
 
+import java.util.Arrays;
+import java.util.Collections;
 import java.util.List;
 import java.util.Objects;
 import java.util.Optional;
 
 public class VertexLabel {
     private static final char NAMESPACE_SEPARATOR = '.';
-    private final String name;
     private final List<String> namespace;
+    private final String name;
 
-    public VertexLabel(String name, String... namespace) {
-        this(name, ImmutableList.copyOf(namespace));
+    public VertexLabel(String outer, String... inner) {
+        if (inner.length <= 0) {
+            this.namespace = Collections.emptyList();
+            this.name = outer;
+        } else {
+            this.namespace = ImmutableList.<String>builder()
+                .add(outer)
+                .add(Arrays.copyOf(inner, inner.length - 1))
+                .build();
+            this.name = inner[inner.length - 1];
+        }
```

Comment:
I think this logic is easier to understand:

    ImmutableList<String> allTerms = ImmutableList.<String>builder().add(outer).add(inner).build();
    int lastPosition = allTerms.size() - 1;
    this.name = allTerms.get(lastPosition);
    this.namespace = allTerms.subList(0, lastPosition);

Reply:
(As pointed out by Wes) I like the use of `Collections.emptyList()` because it just reuses a singleton immutable empty list for `VertexLabel` when there is no namespace. This is good because in most cases users won't specify a namespace.

I think using `subList` to set `namespace` might be a small memory leak because it's just a view of `allTerms` and will still hold onto the last element?

Reply:
That's interesting about the view, although I don't think it's a memory leak, because `allTerms` would get garbage collected when `this.namespace` gets garbage collected, right?

Anyway if you prefer to use `Collections.emptyList()` that's fine.

Another way you could make the logic clearer is to get rid of the `if` by splitting your `then` and `else` into 2 different constructors:

    public VertexLabel(String name) {
        this.namespace = Collections.emptyList();
        this.name = name;
    }

    public VertexLabel(String outer, String... inner) {
        this.namespace = ImmutableList.<String>builder()
            .add(outer)
            .add(Arrays.copyOf(inner, inner.length - 1))
            .build();
        this.name = inner[inner.length - 1];
    }